lib/auth/webauthnwin/api_test.go

```diff
@@ -0,0 +1,255 @@
+)
+
+func TestRegister_errors(t *testing.T) {
+	resetNativeAfterTests(t)
+
+	*webauthnwin.Native = fakeNative{}
+
+	const origin = "https://example.com"
+	okCC := &wanlib.CredentialCreation{
+		Response: protocol.PublicKeyCredentialCreationOptions{
+			Challenge: make([]byte, 32),
+			RelyingParty: protocol.RelyingPartyEntity{
+				ID: "example.com",
+			},
+			Parameters: []protocol.CredentialParameter{
+				{Type: protocol.PublicKeyCredentialType, Algorithm: webauthncose.AlgES256},
+			},
+			AuthenticatorSelection: protocol.AuthenticatorSelection{
+				UserVerification: protocol.VerificationDiscouraged,
+			},
+			Attestation: protocol.PreferNoAttestation,
+		},
+	}
+
+	pwdlessOK := *okCC
+	pwdlessOK.Response.RelyingParty.Name = "Teleport"
+	pwdlessOK.Response.User = protocol.UserEntity{
+		CredentialEntity: protocol.CredentialEntity{
+			Name: "llama",
+		},
+		DisplayName: "Llama",
+		ID:          []byte{1, 2, 3, 4, 5}, // arbitrary
+	}
+	rrk := true
+	pwdlessOK.Response.AuthenticatorSelection.RequireResidentKey = &rrk
+	pwdlessOK.Response.AuthenticatorSelection.UserVerification = protocol.VerificationRequired
+
+	tests := []struct {
+		name     string
+		origin   string
+		createCC func() *wanlib.CredentialCreation
+		wantErr  string
+	}{
+		{
+			// right now there is not need to provide anything in fakeNative
+			name:     "ok",
+			origin:   origin,
+			createCC: func() *wanlib.CredentialCreation { return okCC },
+			wantErr:  "not implemented in fakeNative",
+		},
+		{
+			name:     "nil origin",
+			createCC: func() *wanlib.CredentialCreation { return okCC },
+			wantErr:  "origin",
+		},
+		{
+			name:     "nil cc",
+			origin:   origin,
+			createCC: func() *wanlib.CredentialCreation { return nil },
+			wantErr:  "credential creation required",
+		},
+		{
+			name:   "cc without challenge",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.Challenge = nil
+				return &cp
+			},
+			wantErr: "challenge",
+		},
+		{
+			name:   "cc without RPID",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.RelyingParty.ID = ""
+				return &cp
+			},
+			wantErr: "relying party ID",
+		},
+		{
+			name:   "rrk empty RP name",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.RelyingParty.Name = ""
+				return &cp
+			},
+			wantErr: "relying party name",
+		},
+		{
+			name:   "rrk empty user name",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.User.Name = ""
+				return &cp
+			},
+			wantErr: "user name",
+		},
+		{
+			name:   "rrk empty user display name",
+			origin: origin,
+			createCC: func() *wanlib.CredentialCreation {
+				cp := pwdlessOK
+				cp.Response.User.DisplayName = ""
+				return &cp
+			},
+			wantErr: "user display name",
+		},
+		{
+			name:   "rrk nil user ID",
```

Comment:
Most of the tests here are uninteresting, I think. I'd rather refactor/reuse the validations, have that tested elsewhere, and just make sure we test that the validations are "plugged in" here, plus whatever unique behaviors we have.

Ditto for registration.